# Assignment for Week 2

## Purpose:     
This assignment is intended to help you do the following:   
1.	Modify and add information on the dataset by adding and calculating new columns     
2.	Work with dates in a dataset       
3.	Learn how to write and save your work as a csv file from Jupyter Notebook.       


## Action Items:       
•	Create a new column using the date column.     
  1. Find the maximum date of your data.     
  2. Subtract all the dates on the column from the maximum date.       
  3. Convert the number of days into number of months.       
  4. Save the result on a new column and add to the data set.        
  5. Save the dataset as a csv file.        
  6. Upload the assignment on your GitHub account and Update the README file where you explain everything you did.      

In [0]:
# Import the required Dataset from Franklin GitHub site
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/frankData612/data_612/master/State_Drug_Utilization_Data_2010/State_Drug_Utilization_Data_2010.csv')

In [0]:
# Let's see the first 5 rows of the Dataset
df.head(5)

Unnamed: 0 Utilization Type  ...             Location          NDC
0      128037             MCOU  ...  (31.1801, -91.8749)  37205013705
1      491755             FFSU  ...  (39.3498, -75.5148)  62756076388
2     2568076             FFSU  ...  (44.2853, -99.4632)  46987032311
3      491263             FFSU  ...  (43.3504, -84.5603)    781518501
4      836489             FFSU  ...                  NaN  51552080906

[5 rows x 21 columns]

In [0]:
# Print the Column Names from the 2010 State Drug Utilization Dataset
print(df.columns)

Index(['Unnamed: 0', 'Utilization Type', 'State', 'Labeler Code',
       'Product Code', 'Package Size', 'Year', 'Quarter', 'Product Name',
       'Units Reimbursed', 'Suppression Used', 'Number of Prescriptions',
       'Total Amount Reimbursed', 'Medicaid Amount Reimbursed',
       'Non Medicaid Amount Reimbursed', 'Quarter Begin', 'Quarter Begin Date',
       'Latitude', 'Longitude', 'Location', 'NDC'],
      dtype='object')


In [0]:
# Now let's look at the data types from the 2010 State Drug Utilization Dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156220 entries, 0 to 156219
Data columns (total 21 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Unnamed: 0                      156220 non-null  int64  
 1   Utilization Type                156220 non-null  object 
 2   State                           156220 non-null  object 
 3   Labeler Code                    156220 non-null  int64  
 4   Product Code                    156220 non-null  int64  
 5   Package Size                    156220 non-null  int64  
 6   Year                            156220 non-null  int64  
 7   Quarter                         156220 non-null  int64  
 8   Product Name                    156166 non-null  object 
 9   Units Reimbursed                156220 non-null  float64
 10  Suppression Used                156220 non-null  bool   
 11  Number of Prescriptions         156220 non-null  int64  
 12  Total Amount Rei

We can see from the above output that "Year" and "Quarter" are (int64), but "Quarter Begin" and "Quarter Begin Date" are (object). We will have to convert "Quarter Begin Date" to an (int64) value under a new Column Heading. We also have no "null" values to deal with.

Item 1 - Find the Max and Min dates associated with the 2010 State Drug Utilization Dataset. 

In [0]:
# Let's see the Minimun Date in the "Quarter Begin Date" Column
min(df['Quarter Begin Date'])


'01/01/2010'

In [0]:
# Let's see the Maximun Date in the "Quater Begin Date" Column
max(df['Quarter Begin Date'])

'10/01/2010'

In [0]:
# Let's see the Max Year in our Dataset
max(df['Year'])

2010

In [0]:
# Let's see the Min Year in our Dataset
min(df['Year'])

2010

So, what have we learned? We can see from the syntax above that the dateset is composed of only one year's worth of data (Year - 2010). We also know that the information starts on January 1, 2010 and ends with October 1, 2010 (The Maximum Date).

In [0]:
# Let's see our Data Types again
df.dtypes

Unnamed: 0                                 int64
Utilization Type                          object
Start_Date                        datetime64[ns]
State                                     object
Labeler Code                               int64
Product Code                               int64
Package Size                               int64
Year                                       int64
Quarter                                    int64
Product Name                              object
Units Reimbursed                         float64
Suppression Used                            bool
Number of Prescriptions                    int64
Total Amount Reimbursed                  float64
Medicaid Amount Reimbursed               float64
Non Medicaid Amount Reimbursed           float64
Quarter Begin                             object
Quarter Begin Date                        object
Latitude                                 float64
Longitude                                float64
Location            

In [0]:
# Let's load some additional pandas features
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas.util.testing as tm
from pandas import set_option
import datetime as dt
from dateutil.relativedelta import relativedelta
from datetime import date

This Datset has a "Start Date" of 1 Jan 2010 and an "End Date" of 1 Oct 2010. I am going to insert two new columns into the Dateset called Start_Date and End_Date with static values of 1 Jan 2010 and 1 Oct 2010 respectively. This will allow for the date calculation for this weeks homework assignment. 

In [0]:
# insert the Start Date
df['Start_Date'] = pd.Timestamp('2010-01-01')
df.head(3)



Unnamed: 0 Utilization Type State  ...          NDC   End_Date  Start_Date
0      128037             MCOU    MA  ...  37205013705 2010-10-01  2010-01-01
1      491755             FFSU    FL  ...  62756076388 2010-10-01  2010-01-01
2     2568076             FFSU    UT  ...  46987032311 2010-10-01  2010-01-01

[3 rows x 23 columns]

In [0]:
# Insert the End Date
df['End_Date'] = pd.Timestamp('2010-10-01')
df.head(3)

Unnamed: 0 Utilization Type State  ...          NDC   End_Date  Start_Date
0      128037             MCOU    MA  ...  37205013705 2010-10-01  2010-01-01
1      491755             FFSU    FL  ...  62756076388 2010-10-01  2010-01-01
2     2568076             FFSU    UT  ...  46987032311 2010-10-01  2010-01-01

[3 rows x 23 columns]

Item 2 - Subtract the dates on the Columns from the Maximun date.

Item 3 - Convert the output of the subtraction to months rather than days.

In [0]:
# Now let's take the difference in months between those dates
df['Diff_Months'] = df['End_Date'] - df['Start_Date']
df['Diff_Months'] = df['Diff_Months']/np.timedelta64(1, 'M')
print(df)

        Unnamed: 0 Utilization Type State  ...   End_Date  Start_Date  Diff_Months
0           128037             MCOU    MA  ... 2010-10-01  2010-01-01     8.969383
1           491755             FFSU    FL  ... 2010-10-01  2010-01-01     8.969383
2          2568076             FFSU    UT  ... 2010-10-01  2010-01-01     8.969383
3           491263             FFSU    MN  ... 2010-10-01  2010-01-01     8.969383
4           836489             FFSU    XX  ... 2010-10-01  2010-01-01     8.969383
...            ...              ...   ...  ...        ...         ...          ...
156215     2752214             MCOU    XX  ... 2010-10-01  2010-01-01     8.969383
156216     2519195             MCOU    MI  ... 2010-10-01  2010-01-01     8.969383
156217      815768             MCOU    MD  ... 2010-10-01  2010-01-01     8.969383
156218     2029158             FFSU    DE  ... 2010-10-01  2010-01-01     8.969383
156219     2799238             FFSU    DE  ... 2010-10-01  2010-01-01     8.969383

[15

In [0]:
df.head(3)

Unnamed: 0 Utilization Type State  ...   End_Date  Start_Date  Diff_Months
0      128037             MCOU    MA  ... 2010-10-01  2010-01-01     8.969383
1      491755             FFSU    FL  ... 2010-10-01  2010-01-01     8.969383
2     2568076             FFSU    UT  ... 2010-10-01  2010-01-01     8.969383

[3 rows x 24 columns]

Item 4 - Save the result of the Date Subteraction into a new Column. This was accomplished above syntax with the new Column Heading of "Diff_Months". The value is the same throughout the dataframe when subtracting the End_Date from the Start_Date. The resulting value is: 8.969383 months (Approx. 9 Months)

Item 5 - Save the dataset as a "csv" file.

In [0]:
df.to_csv('df_clean.csv', index=False)

Item 6 - Export Colab BRADLEY_DATA+612_WK2_ASSIGNMENT to GitHub